utils/ami-cleanup/ami_cleanup.py

```diff
@@ -0,0 +1,216 @@
+        del sorted_list[del_offset:]
+    logger.debug(f"Trimmed list:")
+    for ami in sorted_list:
+        logger.debug(f"Date: {ami['CreationDate']}, AMI Name: {ami['Name']}")
+    return sorted_list
+
+
+def check_ami_is_used(ec2_client, asg_client, ami_id):
+    # Check against running instances
+    try:
+        instance_list = ec2_client.describe_instances(
+            Filters=[{"Name": "image-id", "Values": [ami_id]}]
+        )["Reservations"]
+        if len(instance_list) > 0:
+            logger.debug(f"AMI ID: {ami_id} is used by instance(s)")
+            return True
+    except Exception as e:
+        logger.error(f"Failed to get list of Instances using AMI ID {ami_id} : {e}")
+        raise
+
+    # Check against launch configurations
+    try:
+        lc_list = asg_client.describe_launch_configurations()[
+            "LaunchConfigurations"
+        ]
+        for lc in lc_list:
+            if lc["ImageId"] == ami_id:
+                logger.debug(
+                    f"AMI {ami_id} is used by launch configuration {lc['LaunchConfigurationName']}"
+                )
+                return True
+    except Exception as e:
+        logger.error(f"Failed to get list of Launch Configurations using AMI ID {ami_id} : {e}")
+        raise
+
+    # Check against launch templates
+    try:
+        lt_list = ec2_client.describe_launch_templates()["LaunchTemplates"]
+        for lt in lt_list:
+            lt_desc = ec2_client.describe_launch_template_versions(
+                LaunchTemplateId=lt["LaunchTemplateId"],
+                Versions=["$Latest"],
+                Filters=[{"Name": "image-id", "Values": [ami_id]}],
+            )["LaunchTemplateVersions"]
+            if len(lt_desc) > 0:
+                logger.debug(
+                    f"AMI {ami_id} is used by Launch Template {lt['LaunchTemplateName']}"
+                )
+                return True
+    except Exception as e:
+        logger.error(f"Failed to get list of Launch Templates using AMI ID {ami_id} : {e}")
+        raise
+
+    return False
+
+
+def common(lst1, lst2):
+    return list(set(lst1) & set(lst2))
+
+
+def delete_ami(ec2_client, ami_id, dry_run):
+    # Deregister AMI
+
+    try:
+        if not dry_run:
+            ec2_client.deregister_image(ImageId=ami_id)
+            logger.info(f"AMI {ami_id} deregistered")
+        else:
+            logger.info(f"AMI {ami_id} would be deregistered if this wasn't a dry run")
+    except Exception as e:
+        logger.debug(f"Error while deregistering AMI {ami_id}")
+        logger.debug(e)
+
+    # Get a list of snapshots associated with the ami
+    filter_string_1 = snap_filter_1.replace("AMI_ID", ami_id)
+    filter_string_2 = snap_filter_2.replace("AMI_ID", ami_id)
+    try:
+        snapshot_list = ec2_client.describe_snapshots(
+            Filters=[{"Name": "description", "Values": [filter_string_1]}]
+        )["Snapshots"]
+    except Exception as e:
+        logger.debug(e)
```

Comment:
this will continue even if the `describe_snapshots()` fails, maybe need to re-raise exception